x/aswap/handler.go

```diff
@@ -0,0 +1,307 @@
+	day = time.Hour * 24
+	// amount of days for minTimeout
+	timeoutDays = 2
+	minTimeout  = timeoutDays * day
+
+	tagSwapId string = "swap-id"
+	tagAction string = "action"
+)
+
+// RegisterRoutes will instantiate and register
+// all handlers in this package
+func RegisterRoutes(r weave.Registry, auth x.Authenticator, cashctrl cash.Controller) {
+	bucket := NewBucket()
+
+	r.Handle(pathCreateSwap, migration.SchemaMigratingHandler("aswap", CreateSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReleaseSwap, migration.SchemaMigratingHandler("aswap", ReleaseSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReturnReturn, migration.SchemaMigratingHandler("aswap", ReturnSwapHandler{auth, bucket, cashctrl}))
+}
+
+// RegisterQuery will register this bucket as "/aswaps"
+func RegisterQuery(qr weave.QueryRouter) {
+	NewBucket().Register("aswaps", qr)
+}
+
+//---- create
+
+// CreateSwapHandler creates a swap
+type CreateSwapHandler struct {
+	auth   x.Authenticator
+	bucket Bucket
+	bank   cash.CoinMover
+}
+
+var _ weave.Handler = CreateSwapHandler{}
+
+// Check does the validation and sets the cost of the transaction
+func (h CreateSwapHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &weave.CheckResult{
+		GasAllocated: createSwapCost,
+	}
+	return res, nil
+}
+
+// Deliver moves the tokens from sender to the swap account if all conditions are met.
+func (h CreateSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a swap object
+	swap := &Swap{
+		Metadata:     &weave.Metadata{},
+		Src:          msg.Src,
+		Recipient:    msg.Recipient,
+		Timeout:      msg.Timeout,
+		Memo:         msg.Memo,
+		PreimageHash: msg.PreimageHash,
+	}
+	obj, err := h.bucket.Build(db, swap)
+	if err != nil {
+		return nil, err
+	}
+
+	// deposit amounts
+	senderAddr := swap.Src
+	if err := moveCoins(db, h.bank, senderAddr, swap.Address, msg.Amount); err != nil {
+		return nil, err
+	}
+
+	// return id of swap to use in future calls
+	res := &weave.DeliverResult{
+		Tags: []common.KVPair{
+			{Key: []byte(tagSwapId), Value: obj.Key()},
+			{Key: []byte(tagAction), Value: []byte("create-swap")},
+		},
+		Data: obj.Key(),
+	}
+	return res, h.bucket.Save(db, obj)
+}
+
+// validate does all common pre-processing between Check and Deliver.
+func (h CreateSwapHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreateSwapMsg, error) {
+	var msg CreateSwapMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+	if IsExpired(ctx, msg.Timeout.Add(-minTimeout)) {
+		return nil, errors.Wrapf(errors.ErrInvalidInput,
+			"timeout should be a minimum of %d days from now", timeoutDays)
+	}
+	if IsExpired(ctx, msg.Timeout) {
```

Comment:
Why both checks? If the second one will fail, the first one will too, so this is dead code.

Reply:
good catch!